Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+    rw [nonempty_subtype, not_exists] at h
+    exact h
+  have key := ciInf_mem (fun s : {s : Finset X // IsDynCoverOf T F U n s} ↦ (s.val.card : ℕ∞))
+  rw [coverMincard, iInf_subtype'] at k_min
+  rw [← k_min, mem_range, Subtype.exists] at key
+  simp only [ENat.some_eq_coe, Nat.cast_inj, exists_prop] at key
+  exact key
+
+@[simp]
+lemma coverMincard_empty {T : X → X} {U : Set (X × X)} {n : ℕ} : coverMincard T ∅ U n = 0 :=
+  le_antisymm (sInf_le (by simp [IsDynCoverOf])) (zero_le (coverMincard T ∅ U n))
+
+lemma coverMincard_eq_zero_iff (T : X → X) (F : Set X) (U : Set (X × X)) (n : ℕ) :
+    coverMincard T F U n = 0 ↔ F = ∅ := by
+  refine Iff.intro (fun h ↦ subset_empty_iff.1 ?_) (fun F_empt ↦ by rw [F_empt, coverMincard_empty])
+  have := coverMincard_finite_iff T F U n
+  rw [h, eq_true ENat.zero_lt_top, true_iff] at this
+  simp only [IsDynCoverOf, Finset.mem_coe, Nat.cast_eq_zero, Finset.card_eq_zero, exists_eq_right,
+    Finset.not_mem_empty, iUnion_of_empty, iUnion_empty] at this
+  exact this
+
+lemma coverMincard_pos_iff (T : X → X) (F : Set X) (U : Set (X × X)) (n : ℕ) :
+    1 ≤ coverMincard T F U n ↔ F.Nonempty := by
+  rw [ENat.one_le_iff_ne_zero, nonempty_iff_ne_empty, not_iff_not]
+  exact coverMincard_eq_zero_iff T F U n
+
+lemma coverMincard_zero (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    coverMincard T F U 0 = 1 := by
+  apply le_antisymm _ ((coverMincard_pos_iff T F U 0).2 h)
+  rcases h with ⟨x, _⟩
+  have := isDynCoverOf_zero T F U (singleton_nonempty x)
+  rw [← Finset.coe_singleton] at this
+  apply le_of_le_of_eq (coverMincard_le_card this)
+  rw [Finset.card_singleton, Nat.cast_one]
+
+lemma coverMincard_univ (T : X → X) {F : Set X} (h : F.Nonempty) (n : ℕ) :
+    coverMincard T F univ n = 1 := by
+  apply le_antisymm _ ((coverMincard_pos_iff T F univ n).2 h)
+  rcases h with ⟨x, _⟩
+  have := isDynCoverOf_univ T F n (singleton_nonempty x)
+  rw [← Finset.coe_singleton] at this
+  apply le_of_le_of_eq (coverMincard_le_card this)
+  rw [Finset.card_singleton, Nat.cast_one]
+
+lemma coverMincard_iterate_le_pow {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) (m n : ℕ) :
+    coverMincard T F (U ○ U) (m * n) ≤ coverMincard T F U m ^ n := by
+  rcases F.eq_empty_or_nonempty with rfl | F_nonempty
+  · rw [coverMincard_empty]; exact zero_le _
+  rcases n.eq_zero_or_pos with rfl | n_pos
+  · rw [mul_zero, coverMincard_zero T F_nonempty (U ○ U), pow_zero]
+  rcases eq_top_or_lt_top (coverMincard T F U m) with h | h
+  · exact h ▸ le_of_le_of_eq (le_top (α := ℕ∞)) (ENat.top_pow n_pos).symm
+  · rcases (coverMincard_finite_iff T F U m).1 h with ⟨s, s_cover, s_coverMincard⟩
+    rcases s_cover.iterate_le_pow F_inv U_symm n with ⟨t, t_cover, t_le_sn⟩
+    rw [← s_coverMincard]
+    exact le_trans (coverMincard_le_card t_cover) (WithTop.coe_le_coe.2 t_le_sn)
+
+lemma coverMincard_comp_le_pow {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
```

Comment:
```suggestion
lemma coverMincard_le_pow {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
```
There is no composition in the lemma (apart from the entourages, but it's already there in the other lemmas and you never mentioned it).

Reply:
Done, thanks.